File: crypto_service/app/schema/crypto.py
from enum import Enum
from typing import Optional
from pydantic import BaseModel

class UseMode(str, Enum):
    DERIV = "DERIV"
    NORES = "NORES"
    BOTH = "BOTH"
    ENCR = "ENCR"
    DECR = "DECR"
    GEN = "GEN"
    VERIF = "VERIF"
    COMB = "COMB"
    SIGN = "SIGN"

class Algo(str, Enum):
    R2K = "R2K"
    R3K = "R3K"
    R4K = "R4K"
    ECP256 = "ECP256"
    ECP384 = "ECP384"
    ECP512 = "ECP512"
    ECP521 = "ECP521"
    A128 = "A128"
    A192 = "A192"
    A256 = "A256"
    TDES = "TDES"

class KpGenReq(BaseModel):
    algo: Algo
    use_mode: UseMode

class KpGenResp(BaseModel):
    status: str
    pk: Optional[str] = None
    sk_lmk: Optional[str] = None
    class Config:
        from_attributes = True

class SignReq(BaseModel):
    msg: str
    sk_lmk: str
    algo: Algo

class SignResp(BaseModel):
    status: str
    signature: Optional[str] = None
    class Config:
        from_attributes = True
    

class KeyType(str, Enum):
    ZPK = "ZPK"
    PVK = "PVK"
    CVK = "CVK"
    MKAC = "MKAC"
    BDK = "BDK"
    ZMK = "ZMK"
    TMK = "TMK"
    TEK = "TEK"
    DEK = "DEK"
    PEK = "PEK"
    MEK = "MEK"

class EcdhReq(BaseModel):
    eph_pk: str
    algo: Algo
    key_type: KeyType
    use_mode: UseMode
    shared_info: Optional[str] = None

class EcdhResp(BaseModel):
    status: str
    derived_key: Optional[str] = None
    kcv: Optional[str] = None
    recp_eph_pk: Optional[str] = None
    class Config:
        from_attributes = True


class ExpKeyReq(BaseModel):
    key_lmk: str
    kcv: str
    pk: str

class ExpKeyResp(BaseModel):
    status: str
    key_pk: Optional[str] = None
    class Config:
        from_attributes = True


class ExpTr34Req(BaseModel):
    kbpk: str
    kcv: str
    kdh_cert: str
    krd_cert: str
    kdh_sk_lmk: str

class ExpTr34Resp(BaseModel):
    status: str
    aa: Optional[str] = None
    ed: Optional[str] = None
    signature: Optional[str] = None
    class Config:
        from_attributes = True

class RandGenReq(BaseModel):
    len: str

class RandGenResp(BaseModel):
    status: str
    rand_no: Optional[str] = None
    class Config:
        from_attributes = True


class KeyGenReq(BaseModel):
    key_type: KeyType
    use_mode: UseMode
    algo: Algo
    exp_key: Optional[str] = None

class KeyGenResp(BaseModel):
    status: str
    key_lmk: Optional[str] = None
    kcv: Optional[str] = None
    class Config:
        from_attributes = True


class KcvGenReq(BaseModel):
    key_lmk: str

class KcvGenResp(BaseModel):
    status: str
    kcv: Optional[str] = None
    class Config:
        from_attributes = True


class IpekDeriveReq(BaseModel):
    bdk_lmk: str
    iksn: str
    tk: Optional[str] = None
    algo: Algo
    use_mode: UseMode

class IpekDeriveResp(BaseModel):
    status: str
    ipek_lmk: Optional[str] = None
    ipek_tk: Optional[str] = None
    kcv: Optional[str] = None
    class Config:
        from_attributes = True

class EncrMode(str, Enum):
    CBC = "CBC"
    ECB = "ECB"
    CBC_PAD = "CBC_PAD"

class DataEncrReq(BaseModel):
    key_lmk: str
    ksn: Optional[str] = None
    iv: Optional[str] = None
    encr_mode: EncrMode
    msg: str
    algo: Algo

class DataEncrResp(BaseModel):
    status: str
    encr_msg: Optional[str] = None
    class Config:
        from_attributes = True


class DataDecrReq(BaseModel):
    key_lmk: str
    ksn: Optional[str] = None
    iv: Optional[str] = None
    encr_mode: Optional[EncrMode] = None
    encr_msg: str
    algo: Algo

class DataDecrResp(BaseModel):
    status: str
    msg: Optional[str] = None
    class Config:
        from_attributes = True


class MacMode(str, Enum):
    GENERATE = "GENERATE"
    VERIFY = "VERIFY"

class MacReq(BaseModel):
    key_lmk: str
    ksn: Optional[str] = None
    mac_mode: MacMode
    msg: str
    mac: Optional[str] = None

class MacResp(BaseModel):
    status: str
    mac_resp: Optional[str] = None
    class Config:
        from_attributes = True


class TransPinReq(BaseModel):
    key_lmk: str
    ksn: Optional[str] = None
    src_pinblk: str
    dest_key: str
    dest_ksn: Optional[str] = None
    pan: str

class TransPinResp(BaseModel):
    status: str
    dest_pinblk: Optional[str] = None
    class Config:
        from_attributes = True


class ExpTr31Req(BaseModel):
    key_lmk: str
    zmk_lmk: str
    iksn: Optional[str] = None

class ExpTr31Resp(BaseModel):
    status: str
    key_zmk: Optional[str] = None
    class Config:
        from_attributes = True


class WrapReq(BaseModel):
    algo: Algo
    header: Optional[str] = None
    kbpk: str
    key: str

class WrapResp(BaseModel):
    key_kbpk: str
    status: str
    class Config:
        from_attributes = True

class UnwrapReq(BaseModel):
    key_kbpk: str
    kbpk: str

class UnwrapResp(BaseModel):
    key: str
    status: str
    class Config:
        from_attributes = True

File: services/crypto_svc/internal/dto/crypto.go
package dto

// Crypto endpoint request/response models. All enum-ish fields (algo, use_mode,
// key_type, encr_mode, mac_mode) are validated by the service layer, matching the
// Python API which accepts those as string enums.

type KpGenReq struct {
	Algo    string `json:"algo"`
	UseMode string `json:"use_mode"`
}

type KpGenResp struct {
	Status string  `json:"status"`
	Pk     string  `json:"pk,omitempty"`
	SkLmk  string  `json:"sk_lmk,omitempty"`
}

type SignReq struct {
	Msg   string `json:"msg"`
	SkLmk string `json:"sk_lmk"`
	Algo  string `json:"algo"`
}

type SignResp struct {
	Status    string `json:"status"`
	Signature string `json:"signature,omitempty"`
}

type EcdhReq struct {
	EphPk     string `json:"eph_pk"`
	Algo      string `json:"algo"`
	KeyType   string `json:"key_type"`
	UseMode   string `json:"use_mode"`
	SharedInfo string `json:"shared_info,omitempty"`
}

type EcdhResp struct {
	Status     string `json:"status"`
	DerivedKey string `json:"derived_key,omitempty"`
	Kcv        string `json:"kcv,omitempty"`
	RecpEphPk  string `json:"recp_eph_pk,omitempty"`
}

type ExpKeyReq struct {
	KeyLmk string `json:"key_lmk"`
	Kcv    string `json:"kcv"`
	Pk     string `json:"pk"`
}

type ExpKeyResp struct {
	Status string `json:"status"`
	KeyPk  string `json:"key_pk,omitempty"`
}

type ExpTr31Req struct {
	KeyLmk string `json:"key_lmk"`
	ZmkLmk string `json:"zmk_lmk"`
	Iksn   string `json:"iksn,omitempty"`
}

type ExpTr31Resp struct {
	Status string `json:"status"`
	KeyZmk string `json:"key_zmk,omitempty"`
}

type RandGenReq struct {
	Len string `json:"len"`
}

type RandGenResp struct {
	Status string `json:"status"`
	RandNo string `json:"rand_no,omitempty"`
}

type ExpTr34Req struct {
	Kbpk     string `json:"kbpk"`
	Kcv      string `json:"kcv"`
	KdhCert  string `json:"kdh_cert"`
	KrdCert  string `json:"krd_cert"`
	KdhSkLmk string `json:"kdh_sk_lmk"`
}

type ExpTr34Resp struct {
	Status    string `json:"status"`
	Aa        string `json:"aa,omitempty"`
	Ed        string `json:"ed,omitempty"`
	Signature string `json:"signature,omitempty"`
}

type KeyGenReq struct {
	KeyType string `json:"key_type"`
	UseMode string `json:"use_mode"`
	Algo    string `json:"algo"`
	ExpKey  string `json:"exp_key,omitempty"`
}

type KeyGenResp struct {
	Status string `json:"status"`
	KeyLmk string `json:"key_lmk,omitempty"`
	Kcv    string `json:"kcv,omitempty"`
}

type KcvGenReq struct {
	KeyLmk string `json:"key_lmk"`
}

type KcvGenResp struct {
	Status string `json:"status"`
	Kcv    string `json:"kcv,omitempty"`
}

type IpekDeriveReq struct {
	BdkLmk  string `json:"bdk_lmk"`
	Iksn    string `json:"iksn"`
	Tk      string `json:"tk,omitempty"`
	Algo    string `json:"algo"`
	UseMode string `json:"use_mode"`
}

type IpekDeriveResp struct {
	Status  string `json:"status"`
	IpekLmk string `json:"ipek_lmk,omitempty"`
	IpekTk  string `json:"ipek_tk,omitempty"`
	Kcv     string `json:"kcv,omitempty"`
}

type DataEncrReq struct {
	KeyLmk   string `json:"key_lmk"`
	Ksn      string `json:"ksn,omitempty"`
	Iv       string `json:"iv,omitempty"`
	EncrMode string `json:"encr_mode"`
	Msg      string `json:"msg"`
	Algo     string `json:"algo"`
}

type DataEncrResp struct {
	Status  string `json:"status"`
	EncrMsg string `json:"encr_msg,omitempty"`
}

type DataDecrReq struct {
	KeyLmk   string `json:"key_lmk"`
	Ksn      string `json:"ksn,omitempty"`
	Iv       string `json:"iv,omitempty"`
	EncrMode string `json:"encr_mode"`
	EncrMsg  string `json:"encr_msg"`
	Algo     string `json:"algo"`
}

type DataDecrResp struct {
	Status string `json:"status"`
	Msg    string `json:"msg,omitempty"`
}

type MacReq struct {
	KeyLmk  string `json:"key_lmk"`
	Ksn     string `json:"ksn,omitempty"`
	MacMode string `json:"mac_mode"`
	Msg     string `json:"msg"`
	Mac     string `json:"mac,omitempty"`
}

type MacResp struct {
	Status  string `json:"status"`
	MacResp string `json:"mac_resp,omitempty"`
}

type TransPinReq struct {
	KeyLmk   string `json:"key_lmk"`
	Ksn      string `json:"ksn,omitempty"`
	SrcPinblk string `json:"src_pinblk"`
	DestKey  string `json:"dest_key"`
	DestKsn  string `json:"dest_ksn,omitempty"`
	Pan      string `json:"pan"`
}

type TransPinResp struct {
	Status     string `json:"status"`
	DestPinblk string `json:"dest_pinblk,omitempty"`
}

type WrapReq struct {
	Algo   string `json:"algo"`
	Header string `json:"header,omitempty"`
	Kbpk   string `json:"kbpk"`
	Key    string `json:"key"`
}

type WrapResp struct {
	Status   string `json:"status"`
	KeyKbpk  string `json:"key_kbpk"`
}

type UnwrapReq struct {
	KeyKbpk string `json:"key_kbpk"`
	Kbpk    string `json:"kbpk"`
}

type UnwrapResp struct {
	Status string `json:"status"`
	Key    string `json:"key"`
}